### ISSUE TITLE
State map/makePath support?

### ISSUE THREAD

**Issue by falkTX** [brummer10/screcord.lv2#11]
Hi there.
Want to request to add lv2 state path support to this plugin, so the recordings can be integrated in the host state dir instead of in home.
I recently implemented state make/mapPath in carla and mod-host, and zrhythm also is working on it.

(in the case of MOD, v1.10 will have lv2 state support already)

**Comment by brummer10** [comment]
Hi
Okay, I've implemented  LV2_State_Make_Path, but screcord needs this as feature in instantiate, as it didn't save a state on host request, but may save a recorded file at any time. 
Other than that, I noticed that carla provide a empty path when no project-file is loaded, that's not so good. So screcord check if the provided path is empty before using it, otherwise it will use the old ~/lv2record/ path to save files.

**Comment by falkTX** [comment]
Thanks a lot!

> I noticed that carla provide a empty path when no project-file is loaded

DId you try with latest git/develop carla? I return null now if there is no path set.
I see you request a path of "/lv2record" but this is wrong, please use simply "lv2record" instead.

Do you save the file everytime the host calls save? We could link the file to the session by using mapPath and state store/restore, but not sure if truly needed for this plugin

**Comment by brummer10** [comment]
> DId you try with latest git/develop carla? I return null now if there is no path set.

No, just updated and see, "Oh, shit", I use a std::string, that crash when a null pointer return. Anyway I've switch to const char* for safety now, that work.

> Do you save the file everytime the host calls save?

No, screcord open a file (sf_open()) as soon you press "record", it then write chunks to the file. When you press record again it will write the remaining rest and close the file. When the file size extend the max file size, it will close the file and open a new one for continues recording.

**Comment by brummer10** [comment]
> We could link the file to the session by using mapPath and state store/restore, but not sure if truly needed for this plugin

I think I could add that to GxLiveLooper, there it makes sense.

**Comment by falkTX** [comment]
> No, screcord open a file (sf_open()) as soon you press "record", it then write chunks to the file. When you press record again it will write the remaining rest and close the file. When the file size extend the max file size, it will close the file and open a new one for continues recording.

I don't think this will work very well with LV2 state feature.
Did you actually test it? saving and restore, ensuring it is able to load the previous recording?

**Comment by brummer10** [comment]
screcord is a plain recorder, it will never open the file(s) again. 
That was usually the reason why it saves to ~/lv2record. So a user is able to find the records easily, regardless which host is in use, instead searching in the different state directory's of the different hosts. 
I'm still unsure if it makes sense at all to save the files to the state directory.

On GxLiveLooper that is different, GxLiveLooper saves the recorded files when a preset is to be saved (state interface), and open them again when a preset is loaded, so there it could make sense to use map/makePath, as the wave files been part of the session.

**Comment by falkTX** [comment]
ah ok, then I misunderstood the point of the plugin entirely, sorry.
so the use of state path for it is the for the best.

**Comment by brummer10** [comment]
well, maybe I'll add a path selector so that one could select were to save, those could include a (default) option to select the provided path from makePath.

**Comment by falkTX** [comment]
I have been thinking more about this...

I think saving a recording to the session directly is not useful at all.
So I was (very wrong).

It will be very difficult to find for someone who is not aware of how lv2 handles this.
You added a tooltip, but that is covering the issue rather than solving anything, I believe.

I propose the following:
1. Remove support for lv2 state path altogether
2. On the desktop UI, provide a directory file-open-dialog button for choosing where to save to, defaulting to home directory.
This could be saved as part of the lv2 state.
3. For MOD (embed use without UI), we will patch the code to always point it to the directory dedicated for audio recordings.
So when the user finishes recording, they can just go to the file manager area and see a new file created there.

Does this make sense to you?

**Comment by brummer10** [comment]
> Does this make sense to you?

Yes. It's what I think as well. Will do.

Regarding the MOD, there is a zynthian fork here:
https://github.com/brkl/screcord.lv2
which check for /mount/usb0 and record there if it is. I guess that will be nice for the MOD as well, when the mainline kernel is in place. Recording a session on a usb stick would be a nice feature on the MOD.
